src/core/generator/transformations/type_merger.rs

```diff
@@ -0,0 +1,349 @@
+    let total_fields_count =
+        (type_info_1.fields.len() + type_info_2.fields.len()) - matching_fields_count;
+
+    total_fields_count > 0 && (matching_fields_count as f32 / total_fields_count as f32) >= thresh
+}
+
+fn is_type_comparable(type_: &str) -> bool {
+    matches!(
+        type_.to_lowercase().as_str(),
+        "int" | "id" | "boolean" | "float" | "string" | "any" | "empty"
+    )
+}
+
+fn count_matching_fields(
+    config: &Config,
+    type_info_1: &Type,
+    type_info_2: &Type,
+    thresh: f32,
+) -> usize {
+    type_info_1
+        .fields
+        .iter()
+        .filter(|(field_name, field_info_1)| {
+            type_info_2
+                .fields
+                .get(*field_name)
+                .map_or(false, |field_info_2| {
+                    // if they are primitive types then easy to compare else drill down on the inner
+                    // types to check if they satisfies the thresh criteria.
+                    let field1_primitive = is_type_comparable(field_info_1.type_of.as_str());
+                    let field2_primitive = is_type_comparable(field_info_2.type_of.as_str());
+
+                    if field1_primitive && field2_primitive {
+                        field_info_1.type_of == field_info_2.type_of
+                    } else if !field1_primitive && !field2_primitive {
+                        calculate_similarity(
+                            config,
+                            config.types.get(field_info_1.type_of.as_str()).unwrap(),
+                            config.types.get(field_info_2.type_of.as_str()).unwrap(),
+                            thresh,
+                        )
+                    } else {
+                        false
+                    }
+                })
+        })
+        .count()
+}
+
+fn check_for_conflicts(grouped_types: &HashSet<String>, type_name: &str, config: &Config) -> bool {
+    if grouped_types.iter().any(|grp_type_name| {
+        config.types.get(grp_type_name).map_or(false, |type_| {
+            type_
+                .fields
+                .values()
+                .any(|field| field.type_of == type_name)
+        })
+    }) {
+        return true;
+    }
+
+    config.types.get(type_name).map_or(false, |type_| {
+        type_
+            .fields
+            .values()
+            .any(|field| grouped_types.contains(&field.type_of))
+    })
+}
+
+impl super::Transform for TypeMerger {
+    fn apply(&mut self, mut config: Config) -> Config {
+        let mut type_to_merge_type_mapping = HashMap::new();
+        let mut similar_type_group_list: Vec<HashSet<String>> = vec![];
+        let mut visited_types = HashSet::new();
+        let mut i = 0;
+
+        // step 1: identify all the types that satisfies the thresh criteria and group them.
+        for (type_name_1, type_info_1) in config.types.iter() {
+            if visited_types.contains(type_name_1) || type_name_1 == "Query" {
```

Comment:
```suggestion
            if visited_types.contains(type_name_1) || type_name_1 == schema.query {
```

don't hardcode "query"